### PR TITLE
add support for newer C++ -std= flags on Clang/GCC

### PR DESCRIPTION
**Pull request by williamspatrick**
Clang compilers do not support c++23 standard.  Add it and add support for c++26 to both Clang and GCC.

Closes #9053

### CODE REVIEW

**Comment by williamspatrick** [comment]
I've taken the work from @tycho in #11894 but dropped the ObjectiveC parts.  This will enable c++23 on Clang, which is what I care most about.  I expect @tycho can rebase #11894 after this is merged and just handle the ObjectiveC piece.

**Comment by williamspatrick** [comment]
FYI: @eli-schwartz @tristan957

File: mesonbuild/compilers/cpp.py [hunk 1]
```diff
@@ -294,7 +309,11 @@ class ArmLtdClangCPPCompiler(ClangCPPCompiler):
 
 
 class AppleClangCPPCompiler(ClangCPPCompiler):
-    pass
+
+    _CPP23_VERSION = '>=13.0.0'
+    # We don't know which XCode version will include LLVM 17 yet, so use
+    # something absurd.
+    _CPP26_VERSION = '>=99.0.0'
```
Comment (tristan957) [inline mesonbuild/compilers/cpp.py:316]: Wanna prefix this was TODO:

**Review by tristan957 — COMMENTED** [review]
Thanks for your work on this. I will target it for 1.3.0. Unfortunately we missed the 1.2.0 release.

File: mesonbuild/compilers/cpp.py (hunk 1 above)
Comment (williamspatrick) [reply mesonbuild/compilers/cpp.py:316]: Sure.

**Comment by tristan957** [comment]
It is just the project policy that we can't add features after release candidates get made. I think releases that happen more often would be good for the project, but I have no hand in that.

**Comment by williamspatrick** [comment]
> It is just the project policy that we can't add features after release candidates get made. I think releases that happen more often would be good for the project, but I have no hand in that.

Yeah, I understand.  It's just slightly disappointing in that you said yesterday that it needed to be done by Sunday.  I'll end up having to carry patches for 1.1.1 and 1.2 in order to get clang-tidy working on our project (openbmc).

**Comment by tristan957** [comment]
Yeah, that is my bad. I didn't really consider this a "feature". But @eli-schwartz told me it is too late for this. I don't see any reason this would cause a regression between now and Sunday, but it isn't my call :(.

**Comment by tristan957** [comment]
Let's see what Eli says. Thanks for your understanding regardless!

**Comment by tycho** [comment]
I'm kind of annoyed by this whole process because the original bug report I submitted was effectively "I can't use `-std=c++20` with objective-C++ on macOS" and after we figured out the root cause of that, I was invited to submit a pull request. But it turned into this whole debacle of "oh please add C++23 and C++26 since they're missing too" and "oh, we're not accepting it unless you refactor these Clang/GCC classes while you're at it". And now a new form of the change is being submitted as a PR without the Objective-C++ parts I had problems with in the first place!

Why can't we do the simple change to add c++20 to objcpp.py to unbreak my use case?

**Comment by tristan957** [comment]
The speed at which reviews get done in the project is not where it should be, and that is due to lack of reviewers. Please remember that we are all just trying to do the best we can. Unfortunately, I am not really a real reviewer, so I can't merge PRs or anything, just merely provide feedback that I think would be best for the project.

**Comment by williamspatrick** [comment]
Hey @tycho.  I appreciate the work you’ve put into this. I dropped the ObjCpp patch for 3 reasons:

- @tristan957 originally said we could still get into 1.2 if the work was done by Sunday but you said you didn’t have bandwidth to do it by then. I was trying to get the parts I cared about (selfishly) in. 
- I don’t know how to test the ObjCpp stuff as I don’t have any project I work with (and I don’t regularly use a Mac).
- It wasn’t obvious to me if the existing feedback was applicable to the ObjCpp changes.  Specifically: do we need the “detect clang version” stuff in that compiler and should we really put the “c++2a” options or just the mainline ones, since the fallback should pick up “c++2a” as necessary.  I didn’t think we’d get answers to these questions fast enough. 

I think it should be easy to rebase your commit when what I’ve left in is merged. Either you or I can continue those pieces for 1.3 which has a lot more lead time.

**Comment by eli-schwartz** [comment]
> But it turned into this whole debacle of "oh please add C++23 and C++26 since they're missing too"

I took a quick look and I couldn't find anything about that. As far as I'm concerned that's two independent changes you submitted, and if you wanted to cut that down to just one change (the objcpp one) I'd be very supportive of that.

> "oh, we're not accepting it unless you refactor these Clang/GCC classes while you're at it".

I don't care about that, honestly, except to the extent that for any code covered by a PR, I might have review requests about the structure of the new code being added (and that structure may be easiest to accomplish by refactoring a mixin class it inherits from, but it may not require that at all).

I'm a Linux user so I don't know the nuances of the difference between AppleClang objcpp and generic clang objcpp. I guess I wish it was simpler, but what do I know?

> And now a new form of the change is being submitted as a PR without the Objective-C++ parts I had problems with in the first place!

I'll confess to being pretty disappointed about that myself. It's quite an offer to say "okay I can take over this PR for you" and then follow that up by "well no. I just took the part you didn't care about and I did, you can take the part I don't care about and you do". Kind of defeats the original spirit of the offer. 🤔

...

Meanwhile I am just a tiny bit confused because I'd originally thought it was fixed over a year ago, but no, every PR on the topic gets dropped in short order and then people try to scramble and get new features in after release candidate 3, when we are in strict "regressions and crashes" fixing mode while trying to hurry a new release out the door with minimal changes to the freeze base, because suddenly it's an urgent priority (even though it isn't, because it's targeted for 1.3.0).

I guess this urgency is a pretty good reason to run an ad hoc "adopt-a-PR" program while dropping half the contents of the PR being adopted and thereby failing to accomplish the original goals of the original PR author? Apparently?

**Comment by tycho** [comment]
@eli-schwartz For compilers that would allow us to do runtime detection of supported -std= flags, why don't we do that instead of hardcoding lists of standards and compiler versions? It seems a bit counter-productive that we have to change the build system every time a new language standard is added. And the test matrix only becomes more complex over time as well.

**Comment by eli-schwartz** [comment]
> For compilers that would allow us to do runtime detection of supported -std= flags, why don't we do that instead of hardcoding lists of standards and compiler versions? It seems a bit counter-productive that we have to change the build system every time a new language standard is added.

That sounds like something we could plausibly do, although only for clang (which has this output as `note:`s after the error).

**Comment by tycho** [comment]
> > For compilers that would allow us to do runtime detection of supported -std= flags, why don't we do that instead of hardcoding lists of standards and compiler versions? It seems a bit counter-productive that we have to change the build system every time a new language standard is added.
> 
> 
> 
> That sounds like something we could plausibly do, although only for clang (which has this output as `note:`s after the error).

Well, yeah, if we want to spew an ostensibly more helpful error like "only the following standards are supported". But even for GCC isn't it enough to say "-std=foo is not supported" and call it a day? I mean, ultimately this is just an implementation of "compiler.has_argument" which tries to be helpful, but also gets in the way if meson is older than the compiler.

**Comment by eli-schwartz** [comment]
We also show the available stds if you run `meson configure source_dir`.

**Comment by williamspatrick** [comment]
> I guess this urgency is a pretty good reason to run an ad hoc "adopt-a-PR" program while dropping half the contents of the PR being adopted and thereby failing to accomplish the original goals of the original PR author? Apparently?

@eli-schwartz Is there something you feel like I've done wrong here?  Are we somehow limited on the number of PRs we can merge into this repo?  I've already said that:

- The original message said that we could get this in if done by Sunday.
- You aren't clear on what you want for the ObjCpp pieces, so I did what I knew I can handle.

I've left all of @tycho's authorship, SOBs, etc.  I've fixed all of the review comments that you've directly raised, which @tycho said he wasn't going to be able to do in the next 2 weeks.  It seems like I've only made the project better.  Your remarks are kind of off-putting.

I think it would be good if we can get the "design of the ObjCPP flags" figured out in #11894 between you and @tycho.  As I've said, I'm even willing to keep going on that if I get a clear direction.

Can this code please be reviewed as-is to make sure there's no additional feedback from you?

**Comment by tycho** [comment]
I put just the simple objcpp change in a separate PR #11989.

I'd like to later work on making the Clang-specific `-std=` flag support dynamic based on output of commands like these, where the output is human-readable but stable (and could be parsed with regex or something):
```
$ clang++ -dM -xobjective-c++ -E -std=arglbargle /dev/null -o -
error: invalid value 'arglbargle' in '-std=arglbargle'
note: use 'c++98' or 'c++03' for 'ISO C++ 1998 with amendments' standard
note: use 'gnu++98' or 'gnu++03' for 'ISO C++ 1998 with amendments and GNU extensions' standard
note: use 'c++11' for 'ISO C++ 2011 with amendments' standard
[...]
$ clang -dM -xc -E -std=arglbargle /dev/null -o -
note: use 'c89', 'c90', or 'iso9899:1990' for 'ISO C 1990' standard
note: use 'iso9899:199409' for 'ISO C 1990 with amendment 1' standard
note: use 'gnu89' or 'gnu90' for 'ISO C 1990 with GNU extensions' standard
note: use 'c99' or 'iso9899:1999' for 'ISO C 1999' standard
note: use 'gnu99' for 'ISO C 1999 with GNU extensions' standard
[...]
```
etc

**Comment by tristan957** [comment]
@tycho I definitely think that is a great idea. Thanks for taking that up!

**Comment by williamspatrick** [comment]
Is there any holdup that prevents review and merge now?  1.2 was released and some 1.3 content has been merged. Not sure if there is any additional action expected on my part.

**Comment by tristan957** [comment]
@williamspatrick no. All we need is another review from a reviewer.

**Pull request merged**